### ISSUE TITLE
Consider adding a pointer to the request object, from the Batch object

### ISSUE THREAD

**Issue by magicmark** [Yelp/pyramid-hypernova#22]
At Yelp, we want to use https://github.com/Yelp/pyramid-hypernova/issues/17 to forward certain headers from the pyramid server down to the hypernova server (`Cookie`, in this case)

So we want to write a new plugin do this. We _could_ do something like this: (option 1)

```
from pyramid_hypernova.plugins import BasePlugin

class CookieHeaderPropagationPlugin(BasePlugin):
    @staticmethod
    def transform_request_headers(headers):
        current_request = get_current_request()
        cookies = current_request.headers.get('Cookie')
        headers.update({'Cookie', cookies})
        return headers
```

But using this function is [considered bad](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/threadlocal.html#pyramid.threadlocal.get_current_request).

**Other Options:**

2. Create and use a new tween which copies the header from the request object to threading.locals. Read from theading.locals in our plugin instead.

    (This seems to have some of the same the same problems as `get_current_request` in that it's another magic global thing we'd need to mock)

3. Have calls to the `render_react_component` function pass in the request object/headers

    Forcing users to pass in request level stuff every time is cumbersome and a huge breaking change. Internally, we already decided against this, and put the batch object on the request object.

4. When setting up the batch object [here](https://github.com/Yelp/pyramid-hypernova/blob/1487a157a9bbf99b89c1486ee0abd6e9c7a5dc2c/pyramid_hypernova/tweens.py#L16), pass in the request object.

    We would then have "pyramid_request" (or something) be a new attribute on the batch object. This would allow us to have access to the request and the headers inside plugin lifecycle methods, and allow us to implement our new plugin without using `get_current_request`

I'd propose option 4 > 1 > 2 > 3. That said, I'd love to hear if you think there are any better ways of implementing this. Thanks!

cc @chriskuehl @mxmul

**Comment by chriskuehl** [comment]
Option 1 isn't that bad since we're already relying on it in the same package.

I agree that option 4 is the best and don't like 3 or 3.

I feel kinda bad because when I wrote this new plugin interface, I originally wrote it to pass a `request` object in, but then deleted that code when I realized that distributed-context was just relying on global state anyway (so didn't need the `request`). In retrospect I probably should've kept that code for future users.
